File: ~ Learn/Algorithms/Python/Tree/tree.py
class Node:
        def __init__(self, data, left: 'Node'=None, right: 'Node'=None):
            self.data = data
            self.left = left
            self.right = right

        def insert(self, data) -> None:
            if data < self.data:
                if not self.left:
                    self.left = Node(data)
                    return                
                
                self.left.insert(data)
            else:
                if not self.right:
                    self.right = Node(data)
                    return
                
                self.right.insert(data)

class Tree:
    def __init__(self):
        self.root = None


    def insert(self, *data):
        def insertData(data):
            if not self.root:
                self.root = Node(data)

            self.root.insert(data)

        for d in data:
            insertData(d)


if __name__ == '__main__':
    tree = Tree()
    tree.insert(10, 5, 15, 2, 7, 12, 20)

File: ~ Learn/Algorithms/C#/Algorithms/Trees/Tree.cs
using System;
using System.Collections.Generic;

namespace DataStructures.Tree {
    public class Tree {
        public Node Root;

        public class Node {
            public char Name;
            public int Value;

            public Node Left;
            public Node Right;

            public Node(char name, int value) {
                Name = name;
                Value = value;
            }

            public override string ToString() {
                return $"{Name}";
            }
        }

        // Constructor
        public Tree(){}

        // IsEmpty
        public bool IsEmpty() {
            return Root == null;
        }

        public void Add(char name, int value) {
            if ( IsEmpty() ) {
                Root = new Node(name, value);
                return;
            }

            Add(new Node(name, value), Root);

            void Add(Node newNode, Node node) {
                if ( newNode.Value < node.Value ) {
                    if (node.Left == null) {
                        node.Left = newNode;
                    }
                    else {
                        Add(newNode, node.Left);
                    }
                }
                else {
                    if (node.Right == null) {
                        node.Right = newNode;
                    }
                    else {
                        Add(newNode, node.Right);
                    }
                }
            }
        }
    }
}